{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d7358807-3223-428c-91e5-f9b5345a31ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "import os\n",
    "import tempfile\n",
    "import unittest\n",
    "from os import path\n",
    "from pathlib import Path\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "064a1bbd-fec4-4607-a991-27851a506a80",
   "metadata": {},
   "outputs": [],
   "source": [
    "from staramr.blast.JobHandler import JobHandler\n",
    "from staramr.blast.plasmidfinder.PlasmidfinderBlastDatabase import PlasmidfinderBlastDatabase\n",
    "from staramr.blast.pointfinder.PointfinderBlastDatabase import PointfinderBlastDatabase\n",
    "from staramr.blast.resfinder.ResfinderBlastDatabase import ResfinderBlastDatabase\n",
    "from staramr.databases.AMRDatabasesManager import AMRDatabasesManager\n",
    "from staramr.databases.resistance.pointfinder.ARGDrugTablePointfinder import ARGDrugTablePointfinder\n",
    "from staramr.databases.resistance.resfinder.ARGDrugTableResfinder import ARGDrugTableResfinder\n",
    "from staramr.databases.resistance.cge.CGEDrugTableResfinder import CGEDrugTableResfinder\n",
    "from staramr.detection.AMRDetection import AMRDetection\n",
    "from staramr.detection.AMRDetectionResistance import AMRDetectionResistance\n",
    "from staramr.databases.resistance.pointfinder.complex.ComplexMutations import ComplexMutations\n",
    "\n",
    "logger = logging.getLogger('AMRDetectionIT')\n",
    "\n",
    "\n",
    "class AMRDetectionIT(unittest.TestCase):\n",
    "\n",
    "    def setUp(self):\n",
    "        blast_databases_repositories = AMRDatabasesManager.create_default_manager().get_database_repos()\n",
    "        self.resfinder_dir = blast_databases_repositories.get_repo_dir('resfinder')\n",
    "        self.pointfinder_dir = blast_databases_repositories.get_repo_dir('pointfinder')\n",
    "        self.plasmidfinder_dir = blast_databases_repositories.get_repo_dir('plasmidfinder')\n",
    "\n",
    "        self.columns_resfinder = ('Isolate ID', 'Gene', '%Identity', '%Overlap',\n",
    "                                  'HSP Length/Total Length', 'Contig', 'Start', 'End', 'Accession')\n",
    "\n",
    "        self.test_data_dir = path.join(path.dirname(__file__), '..', 'data')\n",
    "        self.resfinder_database = ResfinderBlastDatabase(self.resfinder_dir)\n",
    "        self.resfinder_drug_table = ARGDrugTableResfinder()\n",
    "        self.pointfinder_drug_table = ARGDrugTablePointfinder()\n",
    "        self.cge_drug_table = CGEDrugTableResfinder(self.resfinder_database.get_phenotypes_file())\n",
    "        self.plasmidfinder_database = PlasmidfinderBlastDatabase(self.plasmidfinder_dir)\n",
    "        self.pointfinder_database = None\n",
    "        self.blast_out = tempfile.TemporaryDirectory()\n",
    "        self.blast_handler = JobHandler(\n",
    "            {'resfinder': self.resfinder_database, 'pointfinder': self.pointfinder_database}, 2, self.blast_out.name)\n",
    "\n",
    "        self.outdir = tempfile.TemporaryDirectory()\n",
    "        self.amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                                    self.cge_drug_table, self.blast_handler,\n",
    "                                                    self.pointfinder_drug_table, self.pointfinder_database,\n",
    "                                                    output_dir=self.outdir.name)\n",
    "\n",
    "        self.drug_key_resfinder_invalid_file = path.join(self.test_data_dir, 'gene-drug-tables',\n",
    "                                                         'drug_key_resfinder_invalid.tsv')\n",
    "        self.drug_key_pointfinder_invalid_file = path.join(self.test_data_dir, 'gene-drug-tables',\n",
    "                                                           'drug_key_pointfinder_invalid.tsv')\n",
    "\n",
    "    def tearDown(self):\n",
    "        self.blast_out.cleanup()\n",
    "        self.outdir.cleanup()\n",
    "\n",
    "    def testResfinderAminoglycosideNameSuccess(self):\n",
    "        file = path.join(self.test_data_dir, \"test-aminoglycoside.fsa\")\n",
    "        files = [file]\n",
    "        self.amr_detection.run_amr_detection(files, 99, 90, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        resfinder_results = self.amr_detection.get_resfinder_results()\n",
    "        self.assertEqual(len(resfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = resfinder_results[resfinder_results['Gene'] == \"aac(6')-Iaa\"]\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 100.00, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['Accession'].iloc[0], 'NC_003197', msg='Wrong accession')\n",
    "\n",
    "    def testResfinderExcludeGeneListSuccess(self):\n",
    "        self.amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                                    self.cge_drug_table, self.blast_handler,\n",
    "                                                    self.pointfinder_drug_table, self.pointfinder_database,\n",
    "                                                    output_dir=self.outdir.name,\n",
    "                                                    genes_to_exclude=[\"aac(6')-Iaa_1_NC_003197\"])\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"test-aminoglycoside.fsa\")\n",
    "        files = [file]\n",
    "        self.amr_detection.run_amr_detection(files, 99, 90, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        resfinder_results = self.amr_detection.get_resfinder_results()\n",
    "        self.assertEqual(len(resfinder_results.index), 0, 'Wrong number of rows in result')\n",
    "\n",
    "    def testResFinderCorrectSeq(self):\n",
    "        self.amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                                    self.cge_drug_table, self.blast_handler,\n",
    "                                                    self.pointfinder_drug_table, self.pointfinder_database,\n",
    "                                                    output_dir=self.outdir.name,)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"test-resfinder-correct-seq.fsa\")\n",
    "        files = [file]\n",
    "        self.amr_detection.run_amr_detection(files, 99, 90, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        results = self.amr_detection.get_resfinder_results()\n",
    "        self.assertEqual(len(results.index), 1, 'Wrong number of rows in result')\n",
    "        self.assertEqual(results['Sequence'].iloc[0], \"ATGAGCAAGTTATCTGCATTCTTTATATTTTTGTTTTGCAGCATTGATACCGCAGCAGAGTCTTTGCCAGATTTAAAAATTGAAAAGCTTGATGAAGGCGTTTATGTTCATACTTCGTTTGAAGAAGTTAACAGGTGGGGCGTTGTTCCTAAACATGGTTTGGTGGTTCTTGTAAATGCTGAGGCTTACCTAATTGACACTCCATTTACGGCTAAAGATACTGAAAAGTTAGTCACTTGGTTTGTGGAGCGTGGCTATAAAATAAAAGGCAGCATTTCCTCTCATTTTCATAGCGACAGCACGGGCGGAATAGAGTGGCTTAATTCTCGATCTATCCCCACGTATGCATCTGAATTAACAAATGAACTGCTTAAAAAAGACGGTAAGGTTCAAGCCACAAATTCATTTAGCGGAGTTAACTATTGGCTAGTTAAAAATAAAATTGAAGTTTTTTATCCAGGCCCGGGACACACTCCAGATAACGTAGTGGTTTGGTTGCCTGAAAGGAAAATATTATTCGGTGGTTGTTTTATTAAACCGTACGGTTTAGGCAATTTGGGTGACGCAAATATAGAAGCTTGGCCAAAGTCCGCCAAATTATTAAAGTCCAAATATGGTAAGGCAAAACTGGTTGTTCCAAGTCACAGTGAAGTTGGAGACGCATCACTCTTGAAACTTACATTAGAGCAGGCGGTTAAAGGGTTAAACGAAAGTAAAAAACCATCAAAACCAAGCAACTAA\", \"Incorrect sequence\")\n",
    "\n",
    "    def testNumericalSequenceID(self):\n",
    "        file = path.join(self.test_data_dir, \"test-seq-id.fsa\")\n",
    "        files = [file]\n",
    "        self.amr_detection.run_amr_detection(files, 99, 90, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        resfinder_results = self.amr_detection.get_resfinder_results()\n",
    "        self.assertEqual(len(resfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = resfinder_results[resfinder_results['Gene'] == \"aac(6')-Iaa\"]\n",
    "        self.assertEqual(result['Contig'].iloc[0], \"1\", \"Incorrect contig id\")\n",
    "\n",
    "    def testResfinderBetaLactam2MutationsSuccess(self):\n",
    "        file = path.join(self.test_data_dir, \"beta-lactam-blaIMP-42-mut-2.fsa\")\n",
    "        files = [file]\n",
    "        self.amr_detection.run_amr_detection(files, 99, 90, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        resfinder_results = self.amr_detection.get_resfinder_results()\n",
    "        self.assertEqual(len(resfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = resfinder_results[resfinder_results['Gene'] == 'blaIMP-42']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.73, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '741/741', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0],\n",
    "                         'ampicillin, amoxicillin/clavulanic acid, cefoxitin, ceftriaxone, meropenem',\n",
    "                         'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'].iloc[0],\n",
    "                         'Amoxicillin, Amoxicillin+Clavulanic acid, Ampicillin, Ampicillin+Clavulanic acid, Cefepime, Cefixime, Cefotaxime, Cefoxitin, Ceftazidime, Ertapenem, Imipenem, Meropenem, Piperacillin, Piperacillin+Tazobactam',\n",
    "                         'Wrong phenotype')        \n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'resfinder_beta-lactam-blaIMP-42-mut-2.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['blaIMP-42_1_AB753456'].seq, records['blaIMP-42_1_AB753456'].seq,\n",
    "                         \"records don't match\")\n",
    "\n",
    "    def testResfinderBetaLactam2MutationsSuccessNoPredictedPhenotype(self):\n",
    "        amr_detection = AMRDetection(self.resfinder_database, self.blast_handler, self.pointfinder_database,\n",
    "                                     output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"beta-lactam-blaIMP-42-mut-2.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 99, 90, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        resfinder_results = amr_detection.get_resfinder_results()\n",
    "        self.assertEqual(len(resfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = resfinder_results[resfinder_results['Gene'] == 'blaIMP-42']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.73, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '741/741', msg='Wrong lengths')\n",
    "        self.assertFalse('Predicted Phenotype' in result.columns, 'Should not exist phenotype column')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'resfinder_beta-lactam-blaIMP-42-mut-2.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['blaIMP-42_1_AB753456'].seq, records['blaIMP-42_1_AB753456'].seq,\n",
    "                         \"records don't match\")\n",
    "\n",
    "    def testResfinderBetaLactam2MutationsSuccessNoMatchDrugTable(self):\n",
    "        resfinder_drug_table = ARGDrugTableResfinder(self.drug_key_resfinder_invalid_file)\n",
    "        self.amr_detection = AMRDetectionResistance(self.resfinder_database, resfinder_drug_table,\n",
    "                                                    self.cge_drug_table, self.blast_handler,\n",
    "                                                    self.pointfinder_drug_table, self.pointfinder_database,\n",
    "                                                    output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"beta-lactam-blaIMP-42-mut-2.fsa\")\n",
    "        files = [file]\n",
    "        self.amr_detection.run_amr_detection(files, 99, 90, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        resfinder_results = self.amr_detection.get_resfinder_results()\n",
    "        self.assertEqual(len(resfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = resfinder_results[resfinder_results['Gene'] == 'blaIMP-42']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'unknown[blaIMP-42_1_AB753456]', 'Wrong phenotype')\n",
    "\n",
    "    def testResfinderBetaLactam2MutationsFail(self):\n",
    "        files = [path.join(self.test_data_dir, \"beta-lactam-blaIMP-42-mut-2.fsa\")]\n",
    "        self.amr_detection.run_amr_detection(files, 99.8, 90, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        resfinder_results = self.amr_detection.get_resfinder_results()\n",
    "        self.assertEqual(len(resfinder_results.index), 0, 'Wrong number of rows in result')\n",
    "\n",
    "        self.assertEqual(len(os.listdir(self.outdir.name)), 0, 'File found where none should exist')\n",
    "\n",
    "    def testResfinderBetaLactamDelStartSuccess(self):\n",
    "        file = path.join(self.test_data_dir, \"beta-lactam-blaIMP-42-del-start.fsa\")\n",
    "        files = [file]\n",
    "        self.amr_detection.run_amr_detection(files, 99, 91, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        resfinder_results = self.amr_detection.get_resfinder_results()\n",
    "        self.assertEqual(len(resfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = resfinder_results[resfinder_results['Gene'] == 'blaIMP-42']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 100.00, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 91.90, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '681/741', msg='Wrong lengths')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'resfinder_beta-lactam-blaIMP-42-del-start.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['blaIMP-42_1_AB753456'].seq, records['blaIMP-42_1_AB753456'].seq,\n",
    "                         \"records don't match\")\n",
    "\n",
    "    def testResfinderBetaLactamDelStartFail(self):\n",
    "        files = [path.join(self.test_data_dir, \"beta-lactam-blaIMP-42-del-start.fsa\")]\n",
    "        self.amr_detection.run_amr_detection(files, 99, 92, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        resfinder_results = self.amr_detection.get_resfinder_results()\n",
    "        self.assertEqual(len(resfinder_results.index), 0, 'Wrong number of rows in result')\n",
    "\n",
    "        self.assertEqual(len(os.listdir(self.outdir.name)), 0, 'File found where none should exist')\n",
    "\n",
    "    def testResfinderBetaLactamInsStartSuccess(self):\n",
    "        file = path.join(self.test_data_dir, \"beta-lactam-blaIMP-42-ins-start.fsa\")\n",
    "        files = [file]\n",
    "        self.amr_detection.run_amr_detection(files, 99, 91, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        resfinder_results = self.amr_detection.get_resfinder_results()\n",
    "        self.assertEqual(len(resfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = resfinder_results[resfinder_results['Gene'] == 'blaIMP-42']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.73, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '741/741', msg='Wrong lengths')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'resfinder_beta-lactam-blaIMP-42-ins-start.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(\n",
    "            SeqIO.parse(path.join(self.test_data_dir, 'beta-lactam-blaIMP-42-mut-2.fsa'), 'fasta'))\n",
    "        logger.debug(\"expected_seq=%s\", expected_records['blaIMP-42_1_AB753456'].seq)\n",
    "        logger.debug(\"actual_seq=%s\", records['blaIMP-42_1_AB753456'].seq)\n",
    "        self.assertEqual(expected_records['blaIMP-42_1_AB753456'].seq, records['blaIMP-42_1_AB753456'].seq,\n",
    "                         \"records don't match\")\n",
    "\n",
    "    def testResfinderBetaLactamDelMiddleSuccess(self):\n",
    "        file = path.join(self.test_data_dir, \"beta-lactam-blaIMP-42-del-middle.fsa\")\n",
    "        files = [file]\n",
    "        self.amr_detection.run_amr_detection(files, 99, 91, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        resfinder_results = self.amr_detection.get_resfinder_results()\n",
    "        self.assertEqual(len(resfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = resfinder_results[resfinder_results['Gene'] == 'blaIMP-42']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.33, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '741/741', msg='Wrong lengths')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'resfinder_beta-lactam-blaIMP-42-del-middle.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(\n",
    "            SeqIO.parse(path.join(self.test_data_dir, 'resfinder_beta-lactam-blaIMP-42-del-middle.fsa'), 'fasta'))\n",
    "        logger.debug(\"expected_seq=%s\", expected_records['blaIMP-42_1_AB753456'].seq)\n",
    "        logger.debug(\"actual_seq=%s\", records['blaIMP-42_1_AB753456'].seq)\n",
    "        self.assertEqual(expected_records['blaIMP-42_1_AB753456'].seq, records['blaIMP-42_1_AB753456'].seq,\n",
    "                         \"records don't match\")\n",
    "\n",
    "    def testResfinderBetaLactamDelMiddleReverseComplementSuccess(self):\n",
    "        file = path.join(self.test_data_dir, \"beta-lactam-blaIMP-42-del-middle-rc.fsa\")\n",
    "        files = [file]\n",
    "        self.amr_detection.run_amr_detection(files, 99, 91, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        resfinder_results = self.amr_detection.get_resfinder_results()\n",
    "        self.assertEqual(len(resfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = resfinder_results[resfinder_results['Gene'] == 'blaIMP-42']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.33, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong percent overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '741/741', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0],\n",
    "                         'ampicillin, amoxicillin/clavulanic acid, cefoxitin, ceftriaxone, meropenem',\n",
    "                         'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'].iloc[0],\n",
    "                         'Amoxicillin, Amoxicillin+Clavulanic acid, Ampicillin, Ampicillin+Clavulanic acid, Cefepime, Cefixime, Cefotaxime, Cefoxitin, Ceftazidime, Ertapenem, Imipenem, Meropenem, Piperacillin, Piperacillin+Tazobactam',\n",
    "                         'Wrong phenotype')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'resfinder_beta-lactam-blaIMP-42-del-middle-rc.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(\n",
    "            SeqIO.parse(path.join(self.test_data_dir, 'resfinder_beta-lactam-blaIMP-42-del-middle.fsa'), 'fasta'))\n",
    "        logger.debug(\"expected_seq=%s\", expected_records['blaIMP-42_1_AB753456'].seq)\n",
    "        logger.debug(\"actual_seq=%s\", records['blaIMP-42_1_AB753456'].seq)\n",
    "        self.assertEqual(expected_records['blaIMP-42_1_AB753456'].seq, records['blaIMP-42_1_AB753456'].seq,\n",
    "                         \"records don't match\")\n",
    "\n",
    "    def testResfinderBetaLactamInsMiddleSuccess(self):\n",
    "        file = path.join(self.test_data_dir, \"beta-lactam-blaIMP-42-ins-middle.fsa\")\n",
    "        files = [file]\n",
    "        self.amr_detection.run_amr_detection(files, 97, 99, 99, 90,0,0,0,0,0)\n",
    "\n",
    "        resfinder_results = self.amr_detection.get_resfinder_results()\n",
    "        self.assertEqual(len(resfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = resfinder_results[resfinder_results['Gene'] == 'blaIMP-42']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 98.14, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 101.62, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '753/741', msg='Wrong lengths')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'resfinder_beta-lactam-blaIMP-42-ins-middle.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(\n",
    "            SeqIO.parse(path.join(self.test_data_dir, 'beta-lactam-blaIMP-42-ins-middle.fsa'), 'fasta'))\n",
    "        logger.debug(\"expected_seq=%s\", expected_records['blaIMP-42_1_AB753456'].seq)\n",
    "        logger.debug(\"actual_seq=%s\", records['blaIMP-42_1_AB753456'].seq)\n",
    "        self.assertEqual(expected_records['blaIMP-42_1_AB753456'].seq.upper(), records['blaIMP-42_1_AB753456'].seq,\n",
    "                         \"records don't match\")\n",
    "\n",
    "    def testResfinderBetaLactamTwoCopies(self):\n",
    "        file = path.join(self.test_data_dir, \"beta-lactam-blaIMP-42-mut-2-two-copies.fsa\")\n",
    "        files = [file]\n",
    "        self.amr_detection.run_amr_detection(files, 99, 90, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        resfinder_results = self.amr_detection.get_resfinder_results()\n",
    "        self.assertEqual(len(resfinder_results.index), 2, 'Wrong number of rows in result')\n",
    "\n",
    "        result = resfinder_results.iloc[0]\n",
    "        self.assertEqual(result['Gene'], 'blaIMP-42', 'Wrong gene for result')\n",
    "        self.assertAlmostEqual(result['%Identity'], 99.73, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'], 100.00, places=2, msg='Wrong percent overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'], '741/741', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Contig'], 'blaIMP-42_1_AB753456', msg='Wrong contig name')\n",
    "        self.assertEqual(result['Start'], 61, msg='Wrong start')\n",
    "        self.assertEqual(result['End'], 801, msg='Wrong end')\n",
    "        self.assertEqual(result['Predicted Phenotype'],\n",
    "                         'ampicillin, amoxicillin/clavulanic acid, cefoxitin, ceftriaxone, meropenem',\n",
    "                         'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'],\n",
    "                         'Amoxicillin, Amoxicillin+Clavulanic acid, Ampicillin, Ampicillin+Clavulanic acid, Cefepime, Cefixime, Cefotaxime, Cefoxitin, Ceftazidime, Ertapenem, Imipenem, Meropenem, Piperacillin, Piperacillin+Tazobactam',\n",
    "                         'Wrong phenotype')\n",
    "\n",
    "        result = resfinder_results.iloc[1]\n",
    "        self.assertEqual(result['Gene'], 'blaIMP-42', 'Wrong gene for result')\n",
    "        self.assertAlmostEqual(result['%Identity'], 99.73, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'], 100.00, places=2, msg='Wrong percent overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'], '741/741', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Contig'], 'blaIMP-42_1_AB753456', msg='Wrong contig name')\n",
    "        self.assertEqual(result['Start'], 841, msg='Wrong start')\n",
    "        self.assertEqual(result['End'], 1581, msg='Wrong end')\n",
    "        self.assertEqual(result['Predicted Phenotype'],\n",
    "                         'ampicillin, amoxicillin/clavulanic acid, cefoxitin, ceftriaxone, meropenem',\n",
    "                         'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'],\n",
    "                         'Amoxicillin, Amoxicillin+Clavulanic acid, Ampicillin, Ampicillin+Clavulanic acid, Cefepime, Cefixime, Cefotaxime, Cefoxitin, Ceftazidime, Ertapenem, Imipenem, Meropenem, Piperacillin, Piperacillin+Tazobactam',\n",
    "                         'Wrong phenotype')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'resfinder_beta-lactam-blaIMP-42-mut-2-two-copies.fsa')\n",
    "        records = list(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 2, 'Wrong number of hit records')\n",
    "\n",
    "        expected_file_data = path.join(self.test_data_dir, \"beta-lactam-blaIMP-42-mut-2.fsa\")\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(expected_file_data, 'fasta'))\n",
    "        # verify I get both copies' sequence data\n",
    "        for record in records:\n",
    "            self.assertEqual(expected_records['blaIMP-42_1_AB753456'].seq, record.seq, \"records don't match\")\n",
    "\n",
    "    def testResfinderBetaLactamTwoCopiesOneReverseComplement(self):\n",
    "        file = path.join(self.test_data_dir, \"beta-lactam-blaIMP-42-mut-2-two-copies-one-rev-complement.fsa\")\n",
    "        files = [file]\n",
    "        self.amr_detection.run_amr_detection(files, 99, 90, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        resfinder_results = self.amr_detection.get_resfinder_results()\n",
    "        self.assertEqual(len(resfinder_results.index), 2, 'Wrong number of rows in result')\n",
    "\n",
    "        result = resfinder_results.iloc[0]\n",
    "        self.assertEqual(result['Gene'], 'blaIMP-42', 'Wrong gene for result')\n",
    "        self.assertAlmostEqual(result['%Identity'], 99.73, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'], 100.00, places=2, msg='Wrong percent overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'], '741/741', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Contig'], 'blaIMP-42_1_AB753456', msg='Wrong contig name')\n",
    "        self.assertEqual(result['Start'], 61, msg='Wrong start')\n",
    "        self.assertEqual(result['End'], 801, msg='Wrong end')\n",
    "        self.assertEqual(result['Predicted Phenotype'],\n",
    "                         'ampicillin, amoxicillin/clavulanic acid, cefoxitin, ceftriaxone, meropenem',\n",
    "                         'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'],\n",
    "                         'Amoxicillin, Amoxicillin+Clavulanic acid, Ampicillin, Ampicillin+Clavulanic acid, Cefepime, Cefixime, Cefotaxime, Cefoxitin, Ceftazidime, Ertapenem, Imipenem, Meropenem, Piperacillin, Piperacillin+Tazobactam',\n",
    "                         'Wrong phenotype')\n",
    "\n",
    "        result = resfinder_results.iloc[1]\n",
    "        self.assertEqual(result['Gene'], 'blaIMP-42', 'Wrong gene for result')\n",
    "        self.assertAlmostEqual(result['%Identity'], 99.73, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'], 100.00, places=2, msg='Wrong percent overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'], '741/741', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Contig'], 'blaIMP-42_1_AB753456', msg='Wrong contig name')\n",
    "        self.assertEqual(result['Start'], 1581, msg='Wrong start')\n",
    "        self.assertEqual(result['End'], 841, msg='Wrong end')\n",
    "        self.assertEqual(result['Predicted Phenotype'],\n",
    "                         'ampicillin, amoxicillin/clavulanic acid, cefoxitin, ceftriaxone, meropenem',\n",
    "                         'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'],\n",
    "                         'Amoxicillin, Amoxicillin+Clavulanic acid, Ampicillin, Ampicillin+Clavulanic acid, Cefepime, Cefixime, Cefotaxime, Cefoxitin, Ceftazidime, Ertapenem, Imipenem, Meropenem, Piperacillin, Piperacillin+Tazobactam',\n",
    "                         'Wrong phenotype')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name,\n",
    "                             'resfinder_beta-lactam-blaIMP-42-mut-2-two-copies-one-rev-complement.fsa')\n",
    "        records = list(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 2, 'Wrong number of hit records')\n",
    "\n",
    "        expected_file_data = path.join(self.test_data_dir, \"beta-lactam-blaIMP-42-mut-2.fsa\")\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(expected_file_data, 'fasta'))\n",
    "        # verify I get both copies' sequence data\n",
    "        for record in records:\n",
    "            self.assertEqual(expected_records['blaIMP-42_1_AB753456'].seq, record.seq, \"records don't match\")\n",
    "\n",
    "    def testPointfinderSalmonellaA67PSuccess(self):\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'salmonella')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"gyrA-A67P.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 99, 99, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'gyrA (A67P)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'gyrA-A67P', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'codon', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 67, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'GCC -> CCC (A -> P)', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.96, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '2637/2637', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'ciprofloxacin I/R, nalidixic acid',\n",
    "                         'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'].iloc[0], 'Nalidixic acid,Ciprofloxacin', 'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Notes'].iloc[0], '', msg='The notes do not match.')  # empty string (no notes)\n",
    "        self.assertEqual(result['CGE PMID'].iloc[0], '7492118', msg='The PMIDs do not match.')\n",
    "        self.assertEqual(result['CGE Mechanism'].iloc[0], 'Target modification', msg='The mechanisms do not match.')\n",
    "        self.assertEqual(result['CGE Required Mutation'].iloc[0], 'gyrA_G81C.S.H.D', msg='The required mutation(s) do not match.')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'pointfinder_gyrA-A67P.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['gyrA'].seq.upper(), records['gyrA_1_MH933946.1'].seq.upper(), \"records don't match\")\n",
    "\n",
    "    def testPointfinderSalmonellaS83ISuccess(self):\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'salmonella')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table, \n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"gyrA-S83I.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 99, 99, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'gyrA (S83I)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'gyrA-S83I', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'codon', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 83, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'TCC -> ATC (S -> I)', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.92, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '2637/2637', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'ciprofloxacin I/R, nalidixic acid',\n",
    "                         'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'].iloc[0], 'Nalidixic acid,Ciprofloxacin', 'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE PMID'].iloc[0], '7492118;10471553', msg='The PMIDs do not match.')\n",
    "        self.assertEqual(result['CGE Mechanism'].iloc[0], 'Target modification', msg='The mechanisms do not match.')\n",
    "        self.assertEqual(result['CGE Required Mutation'].iloc[0], 'gyrA_D87K', msg='The required mutation(s) do not match.')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'pointfinder_gyrA-S83I.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['gyrA'].seq.upper(), records['gyrA_1_MH933946.1'].seq.upper(), \"records don't match\")\n",
    "\n",
    "    def testPointfinderSalmonellaA67PSuccessNoPhenotype(self):\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'salmonella')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetection(self.resfinder_database, blast_handler, pointfinder_database,\n",
    "                                     output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"gyrA-A67P.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 99, 99, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'gyrA_1_MH933946.1 (A67P)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'gyrA-A67P', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'codon', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 67, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'GCC -> CCC (A -> P)', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.96, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '2637/2637', msg='Wrong lengths')\n",
    "        self.assertFalse('Predicted Phenotype' in result.columns, 'Should not exist Predicted Phenotype column')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'pointfinder_gyrA-A67P.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['gyrA'].seq.upper(), records['gyrA_1_MH933946.1'].seq.upper(), \"records don't match\")\n",
    "\n",
    "    def testPointfinderSalmonellaA67PDelEndSuccess(self):\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'salmonella')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"gyrA-A67P-del-end.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 99, 99, 98, 90,0,0,0,0,0)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'gyrA (A67P)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'gyrA-A67P-del-end', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'codon', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 67, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'GCC -> CCC (A -> P)', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.96, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 98.22, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '2590/2637', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'ciprofloxacin I/R, nalidixic acid',\n",
    "                         'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'].iloc[0], 'Nalidixic acid,Ciprofloxacin', 'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE PMID'].iloc[0], '7492118', msg='The PMIDs do not match.')\n",
    "        self.assertEqual(result['CGE Mechanism'].iloc[0], 'Target modification', msg='The mechanisms do not match.')\n",
    "        self.assertEqual(result['CGE Required Mutation'].iloc[0], 'gyrA_G81C.S.H.D', msg='The required mutation(s) do not match.')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'pointfinder_gyrA-A67P-del-end.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['gyrA'].seq.upper(), records['gyrA_1_MH933946.1'].seq.upper(), \"records don't match\")\n",
    "\n",
    "    def testPointfinderSalmonellaA67PDelEndFailPlength(self):\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'salmonella')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"gyrA-A67P-del-end.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 99, 99, 99, 90,0,0,0,0,0)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 0, 'Wrong number of rows in result')\n",
    "\n",
    "        self.assertEqual(len(os.listdir(self.outdir.name)), 0, 'File found where none should exist')\n",
    "\n",
    "    def testPointfinderSalmonellaA67PFailPID(self):\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'salmonella')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        files = [path.join(self.test_data_dir, \"gyrA-A67P.fsa\")]\n",
    "        amr_detection.run_amr_detection(files, 99.97, 99, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 0, 'Wrong number of rows in result')\n",
    "\n",
    "        self.assertEqual(len(os.listdir(self.outdir.name)), 0, 'File found where none should exist')\n",
    "\n",
    "    def testPointfinderSalmonellaA67TFail(self):\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'salmonella')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        files = [path.join(self.test_data_dir, \"gyrA-A67T.fsa\")]\n",
    "        amr_detection.run_amr_detection(files, 99, 99, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 0, 'Wrong number of rows in result')\n",
    "\n",
    "        self.assertEqual(len(os.listdir(self.outdir.name)), 0, 'File found where none should exist')\n",
    "\n",
    "    def testPointfinderSalmonellaA67PReverseComplementSuccess(self):\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'salmonella')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"gyrA-A67P-rc.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 99, 99, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        #Make sure that the quality module fails when we pass 0,0,0,0,0 for the quality checking parameters\n",
    "        summary_results = amr_detection.get_summary_results()\n",
    "        self.assertEqual('Failed', summary_results['Quality Module'].iloc[0], 'Quality result not equal')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'gyrA (A67P)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'gyrA-A67P-rc', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'codon', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 67, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'GCC -> CCC (A -> P)', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.96, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '2637/2637', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'ciprofloxacin I/R, nalidixic acid',\n",
    "                         'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'].iloc[0], 'Nalidixic acid,Ciprofloxacin', 'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE PMID'].iloc[0], '7492118', msg='The PMIDs do not match.')\n",
    "        self.assertEqual(result['CGE Mechanism'].iloc[0], 'Target modification', msg='The mechanisms do not match.')\n",
    "        self.assertEqual(result['CGE Required Mutation'].iloc[0], 'gyrA_G81C.S.H.D', msg='The required mutation(s) do not match.')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'pointfinder_gyrA-A67P-rc.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(path.join(self.test_data_dir, \"gyrA-A67P.fsa\"), 'fasta'))\n",
    "        self.assertEqual(expected_records['gyrA'].seq.upper(), records['gyrA_1_MH933946.1'].seq.upper(), \"records don't match\")\n",
    "\n",
    "    def testPointfinderSalmonella_16S_rrSD_C1065T_Success(self):\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'salmonella')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"16S_rrsD-1T1065.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 99, 99, 90, 90,1000,1000000,1000,300,1000)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        #Make sure that the quality module passes when we pass 1000,1000000,1000,300,1000 for the quality checking parameters\n",
    "        summary_results = amr_detection.get_summary_results()\n",
    "        self.assertEqual('Passed', summary_results['Quality Module'].iloc[0], 'Quality result not equal')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == '16S_rrsD (C1065T)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], '16S_rrsD-1T1065', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'nucleotide', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 1065, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'C -> T', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.94, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '1544/1544', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'spectinomycin',\n",
    "                         'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'].iloc[0], 'Spectinomycin', 'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE PMID'].iloc[0], '12402084', msg='The PMIDs do not match.')\n",
    "        self.assertEqual(result['CGE Mechanism'].iloc[0], 'Target modification', msg='The mechanisms do not match.')\n",
    "        self.assertEqual(result['CGE Required Mutation'].iloc[0], '', msg='The required mutation(s) do not match.')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'pointfinder_16S_rrsD-1T1065.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['16S_rrsD'].seq.upper(), records['16S-rrsD_1_CP049983.1'].seq.upper(),\n",
    "                         \"records don't match\")\n",
    "\n",
    "\n",
    "    def testPointfinderEnterococcusFaecium_pbp5_Success(self):\n",
    "        # This test evaluates the correctness of identifying a pbp5 complex mutation.\n",
    "        # That is, several pbp5 mutations that should together confer ampicillin resistence.\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'enterococcus_faecium')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name,\n",
    "                                               complex_mutations= ComplexMutations())\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"pbp5_20.fa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 80, 80, 80, 80, 100, 1000000, 1000, 300, 1000)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "\n",
    "        # 20 point mutations plus 1 (summarized) complex mutation:\n",
    "        self.assertEqual(len(pointfinder_results.index), 21, 'Wrong number of rows in result')\n",
    "\n",
    "        # Test one of the point mutations:\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'pbp5 (A216S)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'pbp5_20', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'codon', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 216, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'GCA -> AGT (A -> S)', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 98.28, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '2037/2037', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'unknown[pbp5 (A216S)]', 'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'].iloc[0], 'Ampicillin', 'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Notes'].iloc[0], \n",
    "                         'The nineteen pbp5 mutations must be present simultaneously for resistance phenotype',\n",
    "                         msg='The notes do not match.')\n",
    "        self.assertEqual(result['CGE PMID'].iloc[0], '25182648', msg='The PMIDs do not match.')\n",
    "        self.assertEqual(result['CGE Mechanism'].iloc[0], 'Target modification', msg='The mechanisms do not match.')\n",
    "        self.assertEqual(result['CGE Required Mutation'].iloc[0], '', msg='The required mutation(s) do not match.')\n",
    "\n",
    "        # Test the complex mutation:\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'pbp5 (A216S), pbp5 (A499T), pbp5 (A68T), pbp5 (D204G), pbp5 (E100Q), pbp5 (E525D), pbp5 (E629V), pbp5 (E85D), pbp5 (G66E), pbp5 (K144Q), pbp5 (L177I), pbp5 (M485A), pbp5 (N496K), pbp5 (P667S), pbp5 (R34Q), pbp5 (S27G), pbp5 (T172A), pbp5 (T324A), pbp5 (V24A), pbp5 (V586L)']\n",
    "\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'pbp5_20', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'complex', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], \"524, 527, 534, 566, 568, 585, 5100, 5144, 5172, 5177, 5204, 5216, 5324, 5485, 5496, 5499, 5525, 5586, 5629, 5667\", msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'complex', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 98.28, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '2037/2037', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'ampicillin', 'Wrong phenotype')\n",
    "        self.assertTrue(pd.isna(result['CGE Predicted Phenotype'].iloc[0]), 'Wrong phenotype')\n",
    "        self.assertTrue(pd.isna(result['CGE Notes'].iloc[0]), msg='The notes do not match.')\n",
    "        self.assertTrue(pd.isna(result['CGE PMID'].iloc[0]), msg='The PMIDs do not match.')\n",
    "        self.assertTrue(pd.isna(result['CGE Mechanism'].iloc[0]), msg='The mechanisms do not match.')\n",
    "        self.assertTrue(pd.isna(result['CGE Required Mutation'].iloc[0]), msg='The required mutation(s) do not match.')\n",
    "\n",
    "    def testPointfinderEnterococcusFaecium_pbp5_Failure(self):\n",
    "        # This test evaluates the correctness of NOT identifying a pbp5 complex mutation.\n",
    "        # The input will be missing one of the mandatory mutations (pbp5 (E629V)).\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'enterococcus_faecium')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name,\n",
    "                                               complex_mutations= ComplexMutations())\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"pbp5_19_failure.fa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 80, 80, 80, 80, 100, 1000000, 1000, 300, 1000)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "\n",
    "        # 19 point mutations, no summarized complex mutation:\n",
    "        self.assertEqual(len(pointfinder_results.index), 19, 'Wrong number of rows in result')\n",
    "\n",
    "        # Test one of the point mutations:\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'pbp5 (A216S)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'pbp5_19_failure', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'codon', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 216, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'GCA -> AGT (A -> S)', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 98.38, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '2037/2037', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'unknown[pbp5 (A216S)]',\n",
    "                         'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'].iloc[0], 'Ampicillin', 'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE PMID'].iloc[0], '25182648', msg='The PMIDs do not match.')\n",
    "        self.assertEqual(result['CGE Mechanism'].iloc[0], 'Target modification', msg='The mechanisms do not match.')\n",
    "        self.assertEqual(result['CGE Required Mutation'].iloc[0], '', msg='The required mutation(s) do not match.')\n",
    "\n",
    "        # Test correct numbers of complex- and codon-type mutations:\n",
    "        self.assertEqual(len(pointfinder_results[pointfinder_results['Type'] == 'codon']), 19, 'Wrong number of codon mutations')\n",
    "        self.assertEqual(len(pointfinder_results[pointfinder_results['Type'] == 'complex']), 0, 'Wrong number of complex mutations')\n",
    "\n",
    "    def testPointfinderEnterococcusFaecium_pbp5_3_m485a_Success(self):\n",
    "        # This test evaluates the correctness of identifying a pbp5 complex mutation\n",
    "        # with only the 3 mandatory mutations (pbp5 (M485A), pbp5 (E629V), pbp5 (P667S)).\n",
    "        # Note that the mutation at position 485 is to an \"A\".\n",
    "        # The complex mutation should confer ampicillin resistence.\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'enterococcus_faecium')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name,\n",
    "                                               complex_mutations= ComplexMutations())\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"pbp5_3_m485a.fa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 80, 80, 80, 80, 100, 1000000, 1000, 300, 1000)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "\n",
    "        # 3 point mutations plus 1 (summarized) complex mutation:\n",
    "        self.assertEqual(len(pointfinder_results.index), 4, 'Wrong number of rows in result')\n",
    "\n",
    "        # Test one of the point mutations:\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'pbp5 (E629V)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'pbp5_3_m485a', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'codon', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 629, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'GAA -> GTT (E -> V)', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.656, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '2037/2037', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'unknown[pbp5 (E629V)]',\n",
    "                         'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'].iloc[0], 'Ampicillin', 'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE PMID'].iloc[0], '25182648', msg='The PMIDs do not match.')\n",
    "        self.assertEqual(result['CGE Mechanism'].iloc[0], 'Target modification', msg='The mechanisms do not match.')\n",
    "        self.assertEqual(result['CGE Required Mutation'].iloc[0], '', msg='The required mutation(s) do not match.')\n",
    "        \n",
    "        # Test the complex mutation:\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'pbp5 (E629V), pbp5 (M485A), pbp5 (P667S)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'pbp5_3_m485a', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'complex', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], \"5485, 5629, 5667\", msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'complex', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.656, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '2037/2037', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'ampicillin',\n",
    "                         'Wrong phenotype')\n",
    "        self.assertTrue(pd.isna(result['CGE Predicted Phenotype'].iloc[0]), 'Wrong phenotype')\n",
    "        self.assertTrue(pd.isna(result['CGE PMID'].iloc[0]), msg='The PMIDs do not match.')\n",
    "        self.assertTrue(pd.isna(result['CGE Mechanism'].iloc[0]), msg='The mechanisms do not match.')\n",
    "        self.assertTrue(pd.isna(result['CGE Required Mutation'].iloc[0]), msg='The required mutation(s) do not match.')\n",
    "\n",
    "    def testPointfinderEnterococcusFaecium_pbp5_3_m485t_Success(self):\n",
    "        # This test evaluates the correctness of identifying a pbp5 complex mutation\n",
    "        # with only the 3 mandatory mutations (pbp5 (M485T), pbp5 (E629V), pbp5 (P667S)).\n",
    "        # Note that the mutation at position 485 is to an \"T\".\n",
    "        # The complex mutation should confer ampicillin resistence.\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'enterococcus_faecium')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name,\n",
    "                                               complex_mutations= ComplexMutations())\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"pbp5_3_m485t.fa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 80, 80, 80, 80, 100, 1000000, 1000, 300, 1000)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "\n",
    "        # 3 point mutations plus 1 (summarized) complex mutation:\n",
    "        self.assertEqual(len(pointfinder_results.index), 4, 'Wrong number of rows in result')\n",
    "\n",
    "        # Test one of the point mutations:\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'pbp5 (E629V)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'pbp5_3_m485t', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'codon', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 629, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'GAA -> GTT (E -> V)', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.705, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '2037/2037', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'unknown[pbp5 (E629V)]',\n",
    "                         'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'].iloc[0], 'Ampicillin', 'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE PMID'].iloc[0], '25182648', msg='The PMIDs do not match.')\n",
    "        self.assertEqual(result['CGE Mechanism'].iloc[0], 'Target modification', msg='The mechanisms do not match.')\n",
    "        self.assertEqual(result['CGE Required Mutation'].iloc[0], '', msg='The required mutation(s) do not match.')\n",
    "        \n",
    "        # Test the complex mutation:\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'pbp5 (E629V), pbp5 (M485T), pbp5 (P667S)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'pbp5_3_m485t', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'complex', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], \"5485, 5629, 5667\", msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'complex', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.705, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '2037/2037', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'ampicillin',\n",
    "                         'Wrong phenotype')\n",
    "        self.assertTrue(pd.isna(result['CGE Predicted Phenotype'].iloc[0]), 'Wrong phenotype')\n",
    "        self.assertTrue(pd.isna(result['CGE PMID'].iloc[0]), msg='The PMIDs do not match.')\n",
    "        self.assertTrue(pd.isna(result['CGE Mechanism'].iloc[0]), msg='The mechanisms do not match.')\n",
    "        self.assertTrue(pd.isna(result['CGE Required Mutation'].iloc[0]), msg='The required mutation(s) do not match.')\n",
    "\n",
    "    def testResfinderPointfinderSalmonella_16S_C1065T_gyrA_A67_beta_lactam_Success(self):\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'salmonella')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"16S_gyrA_beta-lactam.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 99, 99, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        resfinder_results = amr_detection.get_resfinder_results()\n",
    "        self.assertEqual(len(resfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = resfinder_results[resfinder_results['Gene'] == 'blaIMP-42']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.73, places=2, msg='Wrong pid')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0],\n",
    "                         'ampicillin, amoxicillin/clavulanic acid, cefoxitin, ceftriaxone, meropenem',\n",
    "                         'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'].iloc[0],\n",
    "                         'Amoxicillin, Amoxicillin+Clavulanic acid, Ampicillin, Ampicillin+Clavulanic acid, Cefepime, Cefixime, Cefotaxime, Cefoxitin, Ceftazidime, Ertapenem, Imipenem, Meropenem, Piperacillin, Piperacillin+Tazobactam',\n",
    "                         'Wrong phenotype')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'resfinder_16S_gyrA_beta-lactam.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "        expected_records = SeqIO.to_dict(\n",
    "            SeqIO.parse(path.join(self.test_data_dir, 'beta-lactam-blaIMP-42-mut-2.fsa'), 'fasta'))\n",
    "        self.assertEqual(expected_records['blaIMP-42_1_AB753456'].seq.upper(),\n",
    "                         records['blaIMP-42_1_AB753456'].seq.upper(), \"records don't match\")\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 2, 'Wrong number of rows in result')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == '16S_rrsD (C1065T)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], '16S_gyrA_beta-lactam', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'nucleotide', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 1065, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'C -> T', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.94, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '1544/1544', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'spectinomycin',\n",
    "                         'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'].iloc[0], 'Spectinomycin',\n",
    "                         'Wrong phenotype')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'gyrA (A67P)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], '16S_gyrA_beta-lactam', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'codon', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 67, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'GCC -> CCC (A -> P)', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.96, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '2637/2637', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'ciprofloxacin I/R, nalidixic acid',\n",
    "                         'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'].iloc[0], 'Nalidixic acid,Ciprofloxacin',\n",
    "                         'Wrong phenotype')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'pointfinder_16S_gyrA_beta-lactam.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "        self.assertEqual(len(records), 2, 'Wrong number of hit records')\n",
    "        expected_records1 = SeqIO.to_dict(SeqIO.parse(path.join(self.test_data_dir, 'gyrA-A67P.fsa'), 'fasta'))\n",
    "        self.assertEqual(expected_records1['gyrA'].seq.upper(), records['gyrA_1_MH933946.1'].seq.upper(), \"records don't match\")\n",
    "        expected_records2 = SeqIO.to_dict(SeqIO.parse(path.join(self.test_data_dir, '16S_rrsD-1T1065.fsa'), 'fasta'))\n",
    "        self.assertEqual(expected_records2['16S_rrsD'].seq.upper(), records['16S-rrsD_1_CP049983.1'].seq.upper(),\n",
    "                         \"records don't match\")\n",
    "\n",
    "    def testResfinderPointfinderSalmonellaExcludeGenesListSuccess(self):\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'salmonella')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name, genes_to_exclude=['gyrA_1_MH933946.1'])\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"16S_gyrA_beta-lactam.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 99, 99, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        resfinder_results = amr_detection.get_resfinder_results()\n",
    "        self.assertEqual(len(resfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = resfinder_results[resfinder_results['Gene'] == 'blaIMP-42']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == '16S_rrsD (C1065T)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "\n",
    "    def testResfinderPointfinderSalmonella_16Src_C1065T_gyrArc_A67_beta_lactam_Success(self):\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'salmonella')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"16S-rc_gyrA-rc_beta-lactam.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 99, 99, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        resfinder_results = amr_detection.get_resfinder_results()\n",
    "        self.assertEqual(len(resfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = resfinder_results[resfinder_results['Gene'] == 'blaIMP-42']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.73, places=2, msg='Wrong pid')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0],\n",
    "                         'ampicillin, amoxicillin/clavulanic acid, cefoxitin, ceftriaxone, meropenem',\n",
    "                         'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'].iloc[0],\n",
    "                         'Amoxicillin, Amoxicillin+Clavulanic acid, Ampicillin, Ampicillin+Clavulanic acid, Cefepime, Cefixime, Cefotaxime, Cefoxitin, Ceftazidime, Ertapenem, Imipenem, Meropenem, Piperacillin, Piperacillin+Tazobactam',\n",
    "                         'Wrong phenotype')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'resfinder_16S-rc_gyrA-rc_beta-lactam.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "        expected_records = SeqIO.to_dict(\n",
    "            SeqIO.parse(path.join(self.test_data_dir, 'beta-lactam-blaIMP-42-mut-2.fsa'), 'fasta'))\n",
    "        self.assertEqual(expected_records['blaIMP-42_1_AB753456'].seq.upper(),\n",
    "                         records['blaIMP-42_1_AB753456'].seq.upper(), \"records don't match\")\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 2, 'Wrong number of rows in result')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == '16S_rrsD (C1065T)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], '16S-rc_gyrA-rc_beta-lactam', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'nucleotide', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 1065, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'C -> T', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.94, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '1544/1544', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'spectinomycin',\n",
    "                         'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'].iloc[0], 'Spectinomycin', 'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE PMID'].iloc[0], '12402084', msg='The PMIDs do not match.')\n",
    "        self.assertEqual(result['CGE Mechanism'].iloc[0], 'Target modification', msg='The mechanisms do not match.')\n",
    "        self.assertEqual(result['CGE Required Mutation'].iloc[0], '', msg='The required mutation(s) do not match.')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'gyrA (A67P)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], '16S-rc_gyrA-rc_beta-lactam', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'codon', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 67, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'GCC -> CCC (A -> P)', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.96, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '2637/2637', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'ciprofloxacin I/R, nalidixic acid',\n",
    "                         'Wrong phenotype')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'pointfinder_16S-rc_gyrA-rc_beta-lactam.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "        self.assertEqual(len(records), 2, 'Wrong number of hit records')\n",
    "        expected_records1 = SeqIO.to_dict(SeqIO.parse(path.join(self.test_data_dir, 'gyrA-A67P.fsa'), 'fasta'))\n",
    "        self.assertEqual(expected_records1['gyrA'].seq.upper(), records['gyrA_1_MH933946.1'].seq.upper(), \"records don't match\")\n",
    "        expected_records2 = SeqIO.to_dict(SeqIO.parse(path.join(self.test_data_dir, '16S_rrsD-1T1065.fsa'), 'fasta'))\n",
    "        self.assertEqual(expected_records2['16S_rrsD'].seq.upper(), records['16S-rrsD_1_CP049983.1'].seq.upper(),\n",
    "                         \"records don't match\")\n",
    "\n",
    "    def testResfinderExcludeNonMatches(self):\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, self.blast_handler,\n",
    "                                               self.pointfinder_drug_table, self.pointfinder_database,\n",
    "                                               include_negative_results=False, output_dir=self.outdir.name)\n",
    "        file_beta_lactam = path.join(self.test_data_dir, \"beta-lactam-blaIMP-42-mut-2.fsa\")\n",
    "        file_non_match = path.join(self.test_data_dir, \"non-match.fsa\")\n",
    "        files = [file_beta_lactam, file_non_match]\n",
    "        amr_detection.run_amr_detection(files, 99, 90, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        summary_results = amr_detection.get_summary_results()\n",
    "        self.assertEqual(len(summary_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'resfinder_beta-lactam-blaIMP-42-mut-2.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file_beta_lactam, 'fasta'))\n",
    "        self.assertEqual(expected_records['blaIMP-42_1_AB753456'].seq, records['blaIMP-42_1_AB753456'].seq,\n",
    "                         \"records don't match\")\n",
    "\n",
    "    def testResfinderIncludeNonMatches(self):\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, self.blast_handler,\n",
    "                                               self.pointfinder_drug_table, self.pointfinder_database,\n",
    "                                               include_negative_results=True, output_dir=self.outdir.name)\n",
    "        file_beta_lactam = path.join(self.test_data_dir, \"beta-lactam-blaIMP-42-mut-2.fsa\")\n",
    "        file_non_match = path.join(self.test_data_dir, \"non-match.fsa\")\n",
    "        files = [file_beta_lactam, file_non_match]\n",
    "        amr_detection.run_amr_detection(files, 99, 90, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        summary_results = amr_detection.get_summary_results()\n",
    "        self.assertEqual(len(summary_results.index), 2, 'Wrong number of rows in result')\n",
    "\n",
    "        result_beta_lactam = summary_results.loc['beta-lactam-blaIMP-42-mut-2']\n",
    "        self.assertTrue(isinstance(result_beta_lactam, pd.Series), 'Wrong type of results returned')\n",
    "        self.assertEqual(result_beta_lactam['Genotype'], 'blaIMP-42', 'Wrong genotype')\n",
    "\n",
    "        result_sensitive = summary_results.loc['non-match']\n",
    "        self.assertTrue(isinstance(result_sensitive, pd.Series), 'Wrong type of results returned')\n",
    "        self.assertEqual(result_sensitive['Genotype'], 'None', 'Wrong genotype')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'resfinder_beta-lactam-blaIMP-42-mut-2.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file_beta_lactam, 'fasta'))\n",
    "        self.assertEqual(expected_records['blaIMP-42_1_AB753456'].seq, records['blaIMP-42_1_AB753456'].seq,\n",
    "                         \"records don't match\")\n",
    "\n",
    "    def testNonMatches(self):\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, self.blast_handler,\n",
    "                                               self.pointfinder_drug_table, self.pointfinder_database,\n",
    "                                               include_negative_results=True, output_dir=self.outdir.name)\n",
    "        files = [path.join(self.test_data_dir, \"non-match.fsa\")]\n",
    "        amr_detection.run_amr_detection(files, 99, 90, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        summary_results = amr_detection.get_summary_results()\n",
    "        self.assertEqual(len(summary_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result_sensitive = summary_results.loc['non-match']\n",
    "        self.assertTrue(isinstance(result_sensitive, pd.Series), 'Wrong number of results detected')\n",
    "        self.assertEqual(result_sensitive['Genotype'], 'None', 'Wrong genotype')\n",
    "\n",
    "        self.assertEqual(len(os.listdir(self.outdir.name)), 0, 'File found where none should exist')\n",
    "\n",
    "    def testPointfinderCampylobacterA70TSuccess(self):\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'campylobacter')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"gyrA-A70T.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 99, 99, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'gyrA (A70T)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'gyrA-A70T', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'codon', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 70, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'GCC -> ACC (A -> T)', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.96, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '2592/2592', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'ciprofloxacin, nalidixic acid', 'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'].iloc[0], 'Nalidixic acid, Ciprofloxacin', 'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE PMID'].iloc[0], '8384814', msg='The PMIDs do not match.')\n",
    "        self.assertEqual(result['CGE Mechanism'].iloc[0], 'Target modification', msg='The mechanisms do not match.')\n",
    "        self.assertEqual(result['CGE Required Mutation'].iloc[0], '', msg='The required mutation(s) do not match.')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'pointfinder_gyrA-A70T.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['gyrA'].seq.upper(), records['gyrA_1_LR134511.1'].seq.upper(), \"records don't match\")\n",
    "\n",
    "    def testPointfinderCampylobacterA2075GSuccess(self):\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'campylobacter')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"23S-A2075G.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 99, 99, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == '23S (A2075G)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], '23S-A2075G', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'nucleotide', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 2075, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'A -> G', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.97, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '2912/2912', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0],\n",
    "                         'erythromycin, azithromycin, telithromycin, clindamycin', 'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'].iloc[0], 'Azithromycin, Erythromycin, Clindamycin, Telithromycin', 'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE PMID'].iloc[0], '16713726', msg='The PMIDs do not match.')\n",
    "        self.assertEqual(result['CGE Mechanism'].iloc[0], 'Target modification', msg='The mechanisms do not match.')\n",
    "        self.assertEqual(result['CGE Required Mutation'].iloc[0], '', msg='The required mutation(s) do not match.')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'pointfinder_23S-A2075G.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['23S'].seq.upper(), records['23S_1_LR134511.1'].seq.upper(), \"records don't match\")\n",
    "\n",
    "    \"\"\"\n",
    "    def testPointfinderEFaecalisS97NSuccess(self):\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'enterococcus_faecalis')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"gyrA-S97N.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 99, 99, 90, 90)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'gyrA (S97N)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'gyrA-S97N', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'codon', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 97, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'AGT -> AAT (S -> N)', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.96, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '2499/2499', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'unknown[gyrA (S97N)]', 'Wrong phenotype')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'pointfinder_gyrA-S97N.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['gyrA'].seq.upper(), records['gyrA'].seq.upper(), \"records don't match\")\n",
    "\n",
    "    def testPointfinderEFaeciumS83RSuccess(self):\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'enterococcus_faecium')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"gyrA-S83R.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 99, 99, 90, 90)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'gyrA (S83R)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'gyrA-S83R', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'codon', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 83, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'AGT -> CGT (S -> R)', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.96, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '2469/2469', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'unknown[gyrA (S83R)]', 'Wrong phenotype')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'pointfinder_gyrA-S83R.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['gyrA'].seq.upper(), records['gyrA'].seq.upper(), \"records don't match\")\n",
    "\n",
    "    def testPointfinderEFaeciumIns466Success(self):\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'enterococcus_faecium')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"pbp5-ins-466.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 99, 99, 90, 90)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "\n",
    "        self.assertEqual(len(pointfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'pbp5 (ins466S)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'pbp5-ins-466', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'codon', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 466, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], '--T -> AGT (ins -> S)', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.85, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.15, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '2040/2037', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'unknown[pbp5 (ins466S)]', 'Wrong phenotype')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'pointfinder_pbp5-ins-466.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['pbp5'].seq.upper(), records['pbp5'].seq.upper(), \"records don't match\")\n",
    "    \"\"\"\n",
    "\n",
    "    def testPointfinderEscherichiaColiCn42TSuccess(self):\n",
    "        # C-42T (negative coordinate)\n",
    "        # This tests the ability to find a single point mutation in the nucleotide part of a promoter.\n",
    "        # Verified CGE identifies the ampC-promoter (g.-42C>T) mutation.\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'escherichia_coli')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"ampC_promoter_size_53bp-Cn42T.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 99, 99, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'ampC_promoter (C-42T)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'ampC_promoter_size_53bp-Cn42T', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'nucleotide', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], -42, msg='Wrong nucleotide position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'C -> T', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.31, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '145/145', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'ampicillin, amoxicillin/clavulanic acid, cefoxitin',\n",
    "                         'Wrong phenotype')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'pointfinder_ampC_promoter_size_53bp-Cn42T.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['ampC_promoter_size_53bp'].seq.upper(), records['ampC-promoter_1_CP037449.1'].seq.upper(), \"records don't match\")\n",
    "\n",
    "    def testPointfinderMycobacteriumTuberculosisF10ISuccess(self):\n",
    "        # F10I\n",
    "        # This tests the ability to find a single mutation in the codon part of a promoter.\n",
    "        # Verified CGE finds the ahpC-promoter (p.F10I) mutation.\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'mycobacterium_tuberculosis')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"ahpC_promoter_size_180bp-F10I.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 99, 99, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'ahpC_promoter (F10I)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'ahpC_promoter_size_180bp-F10I', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'codon', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 10, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'TTC -> ATC (F -> I)', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.87, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '768/768', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'unknown[ahpC_promoter (F10I)]',\n",
    "                         'Wrong phenotype')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'pointfinder_ahpC_promoter_size_180bp-F10I.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['ahpC_promoter_size_180bp'].seq.upper(), records['ahpC-promoter_1_CP049108.1'].seq.upper(), \"records don't match\")\n",
    "\n",
    "    def testPointfinderEscherichiaColin13insGSuccess(self):\n",
    "        # Insert a G at -13 in the promoter.\n",
    "        # (ampC_promoter_size_53bp / ampC promoter / -13 / - / ins / G)\n",
    "        # This tests the ability to find a single insertion in the nucleotide part of a promoter.\n",
    "        # Verified CGE identifies the ampC-promoter (g.-13_-12insG) mutation.\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'escherichia_coli')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"ampC_promoter_size_53bp-n13insG.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 99, 99, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'ampC_promoter (ins-13G)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'ampC_promoter_size_53bp-n13insG', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'nucleotide', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], -13, msg='Wrong nucleotide position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'ins -> G', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.31, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.69, places=2, msg='Wrong overlap')  # Blast reports 100.69\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '146/145', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'ampicillin, amoxicillin/clavulanic acid, cefoxitin',\n",
    "                         'Wrong phenotype')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'pointfinder_ampC_promoter_size_53bp-n13insG.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['ampC_promoter_size_53bp'].seq.upper(), records['ampC-promoter_1_CP037449.1'].seq.upper(), \"records don't match\")\n",
    "\n",
    "    def testPointfinderEscherichiaColin16insGTSuccess(self):\n",
    "        # Insert a GT at -16 in the promoter.\n",
    "        # (ampC_promoter_size_53bp / ampC promoter / -16 / - / ins / GT)\n",
    "        # This tests the ability to find a double insertion in the nucleotide part of a promoter.\n",
    "        # Verified CGE identifies the ampC-promoter (g.-16_-15insGT) mutation.\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'escherichia_coli')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"ampC_promoter_size_53bp-n16insGT.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 90, 90, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'ampC_promoter (ins-16GT)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'ampC_promoter_size_53bp-n16insGT', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'nucleotide', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], -16, msg='Wrong nucleotide position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'ins -> GT', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 98.639, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 101.38, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '147/145', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'unknown[ampC_promoter (ins-16GT)]',\n",
    "                         'Wrong phenotype')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'pointfinder_ampC_promoter_size_53bp-n16insGT.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['ampC_promoter_size_53bp'].seq.upper(), records['ampC-promoter_1_CP037449.1'].seq.upper(), \"records don't match\")\n",
    "\n",
    "    def testPointfinderNeisseriaGonorrhoeaen57delSuccess(self):\n",
    "        # Delete the A nucleotide at -57 in the promoter.\n",
    "        # (mtrR_promoter_size_66bp / mtrR promoter / -57 / del / A)\n",
    "        # This tests the ability to find a single deletion in the nucleotide part of a promoter.\n",
    "        # Verified CGE identifies the mtrR-promoter (g.-57_-57del) mutation.\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'neisseria_gonorrhoeae')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"mtrR_promoter_size_66bp-n57del.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 90, 90, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'mtrR_promoter (del-57A)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'mtrR_promoter_size_66bp-n57del', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'nucleotide', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], -57, msg='Wrong nucleotide position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'del -> A', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.86, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.0, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '699/699', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'unknown[mtrR_promoter (del-57A)]',\n",
    "                         'Wrong phenotype')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'pointfinder_mtrR_promoter_size_66bp-n57del.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['mtrR_promoter_size_66bp'].seq.upper(), records['mtrR-promoter_1_CP012026.1'].seq.upper().replace('-', ''), \"records don't match\")\n",
    "\n",
    "    def testPointfinderNeisseriaGonorrhoeae78delSuccess(self):\n",
    "        # Delete the GTT/V nucleotides/codon at pos 78 (nucleotide coords) / pos 27 (codon coords)\n",
    "        # Reminder that the Pointfinder database uses 0-base for nucleotides and 1-base for codons.\n",
    "        # DB entry:\n",
    "        # rpsE / rpsE / 78 / - / del / GTT / Spectinomycin / 23183436\n",
    "        # This tests the ability to find a single codon deletion.\n",
    "        # Verified CGE identifies the rpsE codon deletion (rpsE:p.V27_None79del): \n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'neisseria_gonorrhoeae')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"rpsE-del78gtt.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 90, 90, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'rpsE (del27V)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'rpsE-del78gtt', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'codon', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 27, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'GTT -> --- (V -> del)', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.42, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.0, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '519/519', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'unknown[rpsE (del27V)]',\n",
    "                         'Wrong phenotype')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'pointfinder_rpsE-del78gtt.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['rpsE'].seq.upper(), records['rpsE_1_AP023075.1'].seq.upper().replace('-', ''), \"records don't match\")\n",
    "\n",
    "    def testPointfinderEnterococcusFaecium466insSSuccess(self):\n",
    "        # Insert a AGC (S) at position 466 (codon coords).\n",
    "        # DB entry:\n",
    "        # pbp5 / pbp5 / 466 / - / ins / S,D / Ampicillin / 25182648,25182648\n",
    "        # This tests the ability to find a single codon insertion.\n",
    "        # Verified CGE identifies the codon insertion pbp5 (ins466_None467insS). \n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'enterococcus_faecium')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"pbp5-ins465S.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 90, 90, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'pbp5 (ins465S)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'pbp5-ins465S', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'codon', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 465, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], '--- -> AGC (ins -> S)', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.85, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.15, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '2040/2037', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'unknown[pbp5 (ins465S)]',\n",
    "                         'Wrong phenotype')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'pointfinder_pbp5-ins465S.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['pbp5'].seq.upper(), records['pbp5_1_AAK43724.1'].seq.upper().replace('-', ''), \"records don't match\")\n",
    "\n",
    "    def testResfinderCGEPredictedPhenotypes(self):\n",
    "        file = path.join(self.test_data_dir, \"beta-lactam-blaIMP-42-ins-start.fsa\")\n",
    "        files = [file]\n",
    "        self.amr_detection.run_amr_detection(files, 99, 91, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        resfinder_results = self.amr_detection.get_resfinder_results()\n",
    "        self.assertEqual(len(resfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = resfinder_results[resfinder_results['Gene'] == 'blaIMP-42']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0],\n",
    "                         'ampicillin, amoxicillin/clavulanic acid, cefoxitin, ceftriaxone, meropenem',\n",
    "                         msg='Wrong phenotypes.')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'].iloc[0],\n",
    "                         'Amoxicillin, Amoxicillin+Clavulanic acid, Ampicillin, Ampicillin+Clavulanic acid, Cefepime, Cefixime, Cefotaxime, Cefoxitin, Ceftazidime, Ertapenem, Imipenem, Meropenem, Piperacillin, Piperacillin+Tazobactam',\n",
    "                         msg='Wrong CGE-predicted phenotypes.')\n",
    "\n",
    "    def testPointfinderEcoliD87NSuccess(self):\n",
    "        # Specifically tests how to handle duplicate entries in the Pointfinder database.\n",
    "        pointfinder_database = PointfinderBlastDatabase(self.pointfinder_dir, 'escherichia_coli')\n",
    "        blast_handler = JobHandler({'resfinder': self.resfinder_database, 'pointfinder': pointfinder_database}, 2,\n",
    "                                     self.blast_out.name)\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"gyrA-D87N.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files, 99, 99, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        pointfinder_results = amr_detection.get_pointfinder_results()\n",
    "        self.assertEqual(len(pointfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = pointfinder_results[pointfinder_results['Gene'] == 'gyrA (D87N)']\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result.index[0], 'gyrA-D87N', msg='Wrong file')\n",
    "        self.assertEqual(result['Type'].iloc[0], 'codon', msg='Wrong type')\n",
    "        self.assertEqual(result['Position'].iloc[0], 87, msg='Wrong codon position')\n",
    "        self.assertEqual(result['Mutation'].iloc[0], 'GAC -> AAC (D -> N)', msg='Wrong mutation')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 99.96, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '2628/2628', msg='Wrong lengths')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0], 'ciprofloxacin I/R, nalidixic acid',\n",
    "                         'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Predicted Phenotype'].iloc[0], 'Nalidixic acid;Nalidixic acid,Ciprofloxacin', 'Wrong phenotype')\n",
    "        self.assertEqual(result['CGE Notes'].iloc[0], 'D87G or D87Y confer resistance to nalidixic acid only, if occurring alone. Unknown phenotype if D87H occurs alone;D87G or D87Y confer resistance to nalidixic acid only, if occurring alone. Unknown phenotype if D87H occurs alone', msg='The notes do not match.')\n",
    "        self.assertEqual(result['CGE PMID'].iloc[0], '12654733,12654733,12654733,22878251,12654733,1850972;12654733,12654733,12654733,22878251,12654733,1850972', msg='The PMIDs do not match.')\n",
    "        self.assertEqual(result['CGE Mechanism'].iloc[0], 'Target modification;Target modification', msg='The mechanisms do not match.')\n",
    "        self.assertEqual(result['CGE Required Mutation'].iloc[0], ';gyrA_S83L.W.A.V', msg='The required mutation(s) do not match.')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'pointfinder_gyrA-D87N.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['gyrA_1_CP073768.1'].seq.upper(), records['gyrA_1_CP073768.1'].seq.upper(), \"records don't match\")\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "719f02eb-7c9c-4a80-abd3-3b11004f5aa8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
