{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ab61561b-72bd-4a4b-aeea-70ea158ca1de",
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "import tempfile\n",
    "import unittest\n",
    "from os import path\n",
    "\n",
    "\n",
    "from staramr.blast.JobHandler import JobHandler\n",
    "from staramr.blast.plasmidfinder.PlasmidfinderBlastDatabase import PlasmidfinderBlastDatabase\n",
    "from staramr.blast.resfinder.ResfinderBlastDatabase import ResfinderBlastDatabase\n",
    "from staramr.databases.AMRDatabasesManager import AMRDatabasesManager\n",
    "from staramr.databases.resistance.pointfinder.ARGDrugTablePointfinder import ARGDrugTablePointfinder\n",
    "from staramr.databases.resistance.resfinder.ARGDrugTableResfinder import ARGDrugTableResfinder\n",
    "from staramr.databases.resistance.cge.CGEDrugTableResfinder import CGEDrugTableResfinder\n",
    "from staramr.detection.AMRDetectionResistance import AMRDetectionResistance\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0f1833d0-a7dd-4622-a27e-a22989c66bb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "logger = logging.getLogger('AMRDetectionPlasmid')\n",
    "\n",
    "\n",
    "class AMRDetectionPlasmid(unittest.TestCase):\n",
    "\n",
    "    def setUp(self):\n",
    "        blast_databases_repositories = AMRDatabasesManager.create_default_manager().get_database_repos()\n",
    "        self.resfinder_dir = blast_databases_repositories.get_repo_dir(\n",
    "            'resfinder')\n",
    "        self.pointfinder_dir = blast_databases_repositories.get_repo_dir(\n",
    "            'pointfinder')\n",
    "        self.plasmidfinder_dir = blast_databases_repositories.get_repo_dir(\n",
    "            'plasmidfinder')\n",
    "\n",
    "        self.test_data_dir = path.join(path.dirname(__file__), '..', 'data')\n",
    "        self.resfinder_database = ResfinderBlastDatabase(self.resfinder_dir)\n",
    "        self.resfinder_drug_table = ARGDrugTableResfinder()\n",
    "        self.pointfinder_drug_table = ARGDrugTablePointfinder()\n",
    "        self.cge_drug_table = CGEDrugTableResfinder(self.resfinder_database.get_phenotypes_file())\n",
    "        self.plasmidfinder_database = PlasmidfinderBlastDatabase(\n",
    "            self.plasmidfinder_dir)\n",
    "        self.pointfinder_database = None\n",
    "        self.blast_out = tempfile.TemporaryDirectory()\n",
    "        self.blast_handler = JobHandler(\n",
    "            {'resfinder': self.resfinder_database, 'pointfinder': self.pointfinder_database,\n",
    "             'plasmidfinder': self.plasmidfinder_database}, 2, self.blast_out.name)\n",
    "\n",
    "        self.outdir = tempfile.TemporaryDirectory()\n",
    "        self.amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                                    self.cge_drug_table, self.blast_handler,\n",
    "                                                    self.pointfinder_drug_table, self.pointfinder_database,\n",
    "                                                    output_dir=self.outdir.name)\n",
    "\n",
    "    def tearDown(self):\n",
    "        self.blast_out.cleanup()\n",
    "        self.outdir.cleanup()\n",
    "\n",
    "    def testPlasmidfinderNameSuccess(self):\n",
    "        file = path.join(self.test_data_dir, \"test-plasmids-seq.fsa\")\n",
    "        files = [file]\n",
    "        self.amr_detection.run_amr_detection(files, 99, 90, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        plasmidfinder_results = self.amr_detection.get_plasmidfinder_results()\n",
    "        self.assertEqual(len(plasmidfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = plasmidfinder_results[plasmidfinder_results['Plasmid'] == \"IncW\"]\n",
    "\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 100.00, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['Accession'].iloc[0], 'EF633507', msg='Wrong accession')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '243/243', msg='Wrong lengths')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'plasmidfinder_test-plasmids-seq.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['IncW_1__EF633507'].seq, records['IncW_1__EF633507'].seq,\n",
    "                         \"records don't match\")\n",
    "\n",
    "    def testPlasmidfinderGramPositiveSuccess(self):\n",
    "        plasmidfinder_database = PlasmidfinderBlastDatabase(self.plasmidfinder_dir, database_type='gram_positive')\n",
    "        blast_handler = JobHandler(\n",
    "            {'resfinder': self.resfinder_database, 'pointfinder': self.pointfinder_database,\n",
    "             'plasmidfinder': plasmidfinder_database}, 2, self.blast_out.name)\n",
    "        self.outdir = tempfile.TemporaryDirectory()\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, self.pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"test-plasmids-seq.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files=files,\n",
    "                                        pid_threshold=99,\n",
    "                                        plength_threshold_resfinder=90,\n",
    "                                        plength_threshold_pointfinder=90,\n",
    "                                        plength_threshold_plasmidfinder=90,\n",
    "                                        genome_size_lower_bound=0,\n",
    "                                        genome_size_upper_bound=0,\n",
    "                                        minimum_N50_value=0,\n",
    "                                        minimum_contig_length=0,\n",
    "                                        unacceptable_num_contigs=0)\n",
    "\n",
    "        plasmidfinder_results = amr_detection.get_plasmidfinder_results()\n",
    "        self.assertEqual(len(plasmidfinder_results.index), 0, 'Wrong number of rows in result')\n",
    "\n",
    "    def testPlasmidfinderEnterobacterialesSuccess(self):\n",
    "        plasmidfinder_database = PlasmidfinderBlastDatabase(self.plasmidfinder_dir, database_type='enterobacteriales')\n",
    "        blast_handler = JobHandler(\n",
    "            {'resfinder': self.resfinder_database, 'pointfinder': self.pointfinder_database,\n",
    "             'plasmidfinder': plasmidfinder_database}, 2, self.blast_out.name)\n",
    "        self.outdir = tempfile.TemporaryDirectory()\n",
    "        amr_detection = AMRDetectionResistance(self.resfinder_database, self.resfinder_drug_table,\n",
    "                                               self.cge_drug_table, blast_handler,\n",
    "                                               self.pointfinder_drug_table, self.pointfinder_database,\n",
    "                                               output_dir=self.outdir.name)\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"test-plasmids-seq.fsa\")\n",
    "        files = [file]\n",
    "        amr_detection.run_amr_detection(files=files,\n",
    "                                        pid_threshold=99,\n",
    "                                        plength_threshold_resfinder=90,\n",
    "                                        plength_threshold_pointfinder=90,\n",
    "                                        plength_threshold_plasmidfinder=90,\n",
    "                                        genome_size_lower_bound=0,\n",
    "                                        genome_size_upper_bound=0,\n",
    "                                        minimum_N50_value=0,\n",
    "                                        minimum_contig_length=0,\n",
    "                                        unacceptable_num_contigs=0)\n",
    "\n",
    "        plasmidfinder_results = amr_detection.get_plasmidfinder_results()\n",
    "        self.assertEqual(len(plasmidfinder_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = plasmidfinder_results[plasmidfinder_results['Plasmid'] == \"IncW\"]\n",
    "\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertAlmostEqual(result['%Identity'].iloc[0], 100.00, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(result['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(result['Accession'].iloc[0], 'EF633507', msg='Wrong accession')\n",
    "        self.assertEqual(result['HSP Length/Total Length'].iloc[0], '243/243', msg='Wrong lengths')\n",
    "\n",
    "        hit_file = path.join(self.outdir.name, 'plasmidfinder_test-plasmids-seq.fsa')\n",
    "        records = SeqIO.to_dict(SeqIO.parse(hit_file, 'fasta'))\n",
    "\n",
    "        self.assertEqual(len(records), 1, 'Wrong number of hit records')\n",
    "\n",
    "        expected_records = SeqIO.to_dict(SeqIO.parse(file, 'fasta'))\n",
    "        self.assertEqual(expected_records['IncW_1__EF633507'].seq, records['IncW_1__EF633507'].seq,\n",
    "                         \"records don't match\")\n",
    "\n",
    "    def testDetailedSummary_ResPlasmid(self):\n",
    "        file = path.join(self.test_data_dir, \"test-detailed-summary.fsa\")\n",
    "        files = [file]\n",
    "        self.amr_detection.run_amr_detection(files, 99, 90, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        detailed_summary_results = self.amr_detection.get_detailed_summary_results()\n",
    "        self.assertEqual(len(detailed_summary_results.index), 3, 'Wrong number of rows in result')\n",
    "\n",
    "        plasmid_type = detailed_summary_results[detailed_summary_results['Data'] == \"rep1\"]\n",
    "        self.assertEqual(len(plasmid_type.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(plasmid_type['Predicted Phenotype'].iloc[0], '', msg='Wrong predicted phenotype')\n",
    "        self.assertAlmostEqual(plasmid_type['%Identity'].iloc[0], 100.00, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(plasmid_type['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(plasmid_type['Accession'].iloc[0], 'AY357120', msg='Wrong accession')\n",
    "        self.assertEqual(plasmid_type['HSP Length/Total Length'].iloc[0], '1491/1491', msg='Wrong lengths')\n",
    "        self.assertEqual(plasmid_type['Data Type'].iloc[0], 'Plasmid', msg='Wrong data type')\n",
    "\n",
    "        res_type = detailed_summary_results[detailed_summary_results['Data'] == \"tet(47)\"]\n",
    "        self.assertEqual(len(res_type.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(res_type['Predicted Phenotype'].iloc[0], 'tetracycline', msg='Wrong predicted phenotype')\n",
    "        self.assertAlmostEqual(res_type['%Identity'].iloc[0], 100.00, places=2, msg='Wrong pid')\n",
    "        self.assertAlmostEqual(res_type['%Overlap'].iloc[0], 100.00, places=2, msg='Wrong overlap')\n",
    "        self.assertEqual(res_type['Accession'].iloc[0], 'KR857681', msg='Wrong accession')\n",
    "        self.assertEqual(res_type['HSP Length/Total Length'].iloc[0], '1248/1248', msg='Wrong lengths')\n",
    "        self.assertEqual(res_type['Data Type'].iloc[0], 'Resistance', msg='Wrong data type')\n",
    "\n",
    "    def testResistancePlasmidGenesSummary(self):\n",
    "        file = path.join(self.test_data_dir, \"test-resistance-plasmid.fsa\")\n",
    "        files = [file]\n",
    "        self.amr_detection.run_amr_detection(files, 99, 90, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        summary_results = self.amr_detection.get_summary_results()\n",
    "\n",
    "        self.assertEqual(len(summary_results.index), 1, 'Wrong number of rows in result')\n",
    "\n",
    "        result = summary_results[summary_results['Genotype'] == \"blaIMP-42\"]\n",
    "        self.assertEqual(len(result.index), 1, 'Wrong number of results detected')\n",
    "        self.assertEqual(result['Predicted Phenotype'].iloc[0],\n",
    "                         'ampicillin, amoxicillin/clavulanic acid, cefoxitin, ceftriaxone, meropenem',\n",
    "                         msg='Wrong Predicted Phenotype')\n",
    "        self.assertEqual(result['Plasmid'].iloc[0], 'IncW', msg='Wrong Plasmid Type')\n",
    "\n",
    "    def testIndexRangePlasmids(self):\n",
    "        file = path.join(self.test_data_dir, \"test-index-range-plasmid.fsa\")\n",
    "        files = [file]\n",
    "        self.amr_detection.run_amr_detection(files, 99, 90, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        summary_results = self.amr_detection.get_summary_results()\n",
    "\n",
    "        self.assertEqual(len(summary_results.index), 1, 'Wrong number of rows')\n",
    "\n",
    "        self.assertEqual(summary_results['Genotype'].iloc[0], 'None', msg='Wrong Genotype value')\n",
    "        self.assertEqual(summary_results['Predicted Phenotype'].iloc[0], 'Susceptible',\n",
    "                         msg='Wrong Predicted Phenotype value')\n",
    "        self.assertEqual(summary_results['Plasmid'].iloc[0], 'IncFII(pKPX1)', msg='Wrong Plasmid Type')\n",
    "\n",
    "    def testParseUnderscoresBracketsInFASTA(self):\n",
    "        # Tests to ensure that the plasmid AMR detection can properly parse FASTA record IDs that have\n",
    "        # underscores within brackets. For example: \"rep21_24_rep(CN1_plasmid2)_NC_022227\". The \"rep(CN1_plasmid2)\"\n",
    "        # needs to be parsed as a single element.\n",
    "\n",
    "        file = path.join(self.test_data_dir, \"plasmid-underscore-brackets.fsa\")\n",
    "        files = [file]\n",
    "        self.amr_detection.run_amr_detection(files, 99, 90, 90, 90,0,0,0,0,0)\n",
    "\n",
    "        summary_results = self.amr_detection.get_summary_results()\n",
    "\n",
    "        self.assertEqual(len(summary_results.index), 1, 'Wrong number of rows')\n",
    "\n",
    "        self.assertEqual(summary_results['Genotype'].iloc[0], 'None', msg='Wrong Genotype value')\n",
    "        self.assertEqual(summary_results['Predicted Phenotype'].iloc[0], 'Susceptible',\n",
    "                         msg='Wrong Predicted Phenotype value')\n",
    "        self.assertEqual(summary_results['Plasmid'].iloc[0], 'rep21', msg='Wrong Plasmid Type')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba130189-2932-4605-9a7c-5f3946116f46",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
